### PR TITLE
[WIP] fixes bug 1321873 - Make e2e suite work with PhantomJS

### PR DESCRIPTION
**Pull request by peterbe**
@m8ttyB r?

I've added you to have write permissions on this project. That way I should be able to assign you too.

@adngdb Would be nice if you can test this too. 

```
py.test --tb=native  --driver PhantomJS --base-url https://crash-stats.allizom.org  --html /tmp/results/index.html tests/
```

### CODE REVIEW

**Comment by willkg** [comment]
Grabbing this to review.

**Comment by willkg** [comment]
I installed Firefox 44.0.2 and ran the tests using the master branch against -stage and everything passes.

I then ran the tests using this branch against -stage and hit this issue:

```
__________________________________ TestSuperSearch.test_search_with_multiple_lines __________________________________

self = <tests.test_search.TestSuperSearch instance at 0x7f6f204ea998>, base_url = 'https://crash-stats.allizom.org'
selenium = <selenium.webdriver.firefox.webdriver.WebDriver (session="cf86b2c2-19c2-4c76-9a49-ca6fe2e081b0")>

    @pytest.mark.nondestructive
    def test_search_with_multiple_lines(self, base_url, selenium):
        csp = CrashStatsHomePage(selenium, base_url).open()
        cs_super = csp.header.click_super_search()
        # Do an advanced search
        cs_super.click_new_line()
>       cs_super.select_facet('0', 'release channel')

tests/test_search.py:40: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pages/super_search_page.py:95: in select_facet
    'waiting for %s to be clickable' % (xpath)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <selenium.webdriver.support.wait.WebDriverWait (session="cf86b2c2-19c2-4c76-9a49-ca6fe2e081b0")>
method = <selenium.webdriver.support.expected_conditions.element_to_be_clickable object at 0x7f6f20556350>
message = 'waiting for //div[@class="select2-result-label" and text()[contains(.,"release channel")]] to be clickable'

    def until(self, method, message=''):
        """Calls the method provided with the driver as an argument until the \
            return value is not False."""
        screen = None
        stacktrace = None
    
        end_time = time.time() + self._timeout
        while True:
            try:
                value = method(self._driver)
                if value:
                    return value
            except self._ignored_exceptions as exc:
                screen = getattr(exc, 'screen', None)
                stacktrace = getattr(exc, 'stacktrace', None)
            time.sleep(self._poll)
            if time.time() > end_time:
                break
>       raise TimeoutException(message, screen, stacktrace)
E       TimeoutException: Message: waiting for //div[@class="select2-result-label" and text()[contains(.,"release channel")]] to be clickable
E       Stacktrace:
E           at FirefoxDriver.prototype.findElementInternal_ (file:///tmp/tmpLeHX03/extensions/fxdriver@googlecode.com/components/driver-component.js:10770)
E           at FirefoxDriver.prototype.findElement (file:///tmp/tmpLeHX03/extensions/fxdriver@googlecode.com/components/driver-component.js:10779)
E           at DelayedCommand.prototype.executeInternal_/h (file:///tmp/tmpLeHX03/extensions/fxdriver@googlecode.com/components/command-processor.js:12661)
E           at DelayedCommand.prototype.executeInternal_ (file:///tmp/tmpLeHX03/extensions/fxdriver@googlecode.com/components/command-processor.js:12666)
E           at DelayedCommand.prototype.execute/< (file:///tmp/tmpLeHX03/extensions/fxdriver@googlecode.com/components/command-processor.js:12608)

../../../../.virtualenvs/socorro-e2e/local/lib/python2.7/site-packages/selenium/webdriver/support/wait.py:80: TimeoutException
-------------------------------------------------- pytest-selenium --------------------------------------------------
URL: https://crash-stats.allizom.org/search/?product=Firefox&_dont_run=1
```

I run with `-x` so it quits after the first failure. 24 tests passed, so there's at least one failure and possibly more.

I haven't gotten phantomjs installed on my machine, so I haven't tested that side of things, yet.

I think I can push to @peterbe's branch, so I'll work on fixing that failure.

**Comment by adngdb** [comment]
I got PhantomJS working: download from the [website](http://phantomjs.org/download.html), extract and put the binary file in there in your ``PATH``. That worked for me. 

However, one of the tests is hanging: ``tests/test_crash_reports.py:42``. I'll revisit this after you fix the test failing with Firefox, Will.

**Comment by peterbe** [comment]
I discovered, when I tried to start using phantom that `cs_super.select_facet('0', 'release channel')` never worked when using Firefox. Phantom got a little further but failed. So I hacked on it, using phantomjs, till it worked and actually performed a supersearch query filtered by release channel.
However, it's possible I never checked back on using Firefox again. I think it's worth pursuing to try to get this to work in both Firefox and Phantomjs.

**Comment by willkg** [comment]
@peterbe Good to know.

I can run the tests with phantomjs and firefox 44 now, so I've got all the bits I need to work on this. Working on it today.

File: e2e-tests/pages/super_search_page.py [hunk 1]
```diff
@@ -28,10 +29,10 @@ class CrashStatsSuperSearch(CrashStatsBasePage):
     _new_line_button_locator = (By.CSS_SELECTOR, 'button.new-line')
     _highlighted_text_locator = (By.CSS_SELECTOR, 'li[class*="highlighted"]')
     _facet_text_locator = (By.CSS_SELECTOR, '#advanced-search fieldset[id="%s"] > div:nth-child(2) span:first-child')
-    _facet_field_locator = (By.CSS_SELECTOR, '#advanced-search fieldset[id="%s"] .field')
+    _facet_field_locator = (By.CSS_SELECTOR, '#advanced-search fieldset[id="%s"] select.field')
```
Comment (willkg) [inline e2e-tests/pages/super_search_page.py]: I'm pretty sure that's not the thing we want to be using. That's the hidden select, but I think we want to be either targeting the `<a>` with the javascript or the `<input>`.

File: e2e-tests/pages/super_search_page.py [hunk 2]
```diff
@@ -80,26 +81,43 @@ def click_new_line(self):
...
     def select_facet(self, line_id, field):
         input_locator = (self._facet_field_locator[0], self._facet_field_locator[1] % line_id)
-        self.wait.until(lambda s: self.is_element_displayed(*self._highlighted_text_locator))
         facet_field = self.find_element(*input_locator)
-        facet_field.send_keys(field)
-        self.wait.until(lambda s: self.is_element_displayed(*self._highlighted_text_locator))
-        facet_field.send_keys(Keys.RETURN)
+        facet_field.click()
+        xpath = '//div[@class="select2-result-label" and' +\
+            ' text()[contains(.,"%s")]]' % (field)
```
Comment (willkg) [inline e2e-tests/pages/super_search_page.py]: I have two problems with this change. First is that we don't use XPATH in any of the e2e tests--we use CSS selectors everywhere. Second is that this code isn't selecting the facet when run with the Firefox webdriver.

If we want this change, then we need to either convert it to a CSS selector or provide a comment about why that's impossible. If it turns out it's impossible, I suggest we instead spend some time making the HTML/JS/CSS code that makes up supersearch easier to write tests for. Right now it looks like a bunch of autogenerated mumbo-jumbo that's really hard to key off of.

**Comment by willkg** [comment]
I spent the morning working on this.

First, I don't see `.select_facet()` failing when running the e2e tests in the master branch when running with Firefox. I slowed down the tests to verify that they're doing the right thing and compared it to what I'm seeing on allizom and it all looks fine to me. Is there a bug for this issue? What made you think it never worked?

Second, switching to XPATH seems like a bad idea. We use CSS selectors everywhere else, so this goes down a path that's not in line with how these tests are written. In order to land this PR, I think we'd need to re-re-do those changes.

Third, even if we did the work to finish up this PR and land it, we're adding an ongoing maintenance burden of supporting *both* the Firefox and PhantomJS webdrivers. I mulled over this for a while and tossed my thoughts in https://bugzilla.mozilla.org/show_bug.cgi?id=1321873#c2 . Bottom line is that I don't think this is worth it and we should abandon ship at this point.

Given all that, I'm going to close this out. If there's a compelling reason to add support for PhantomJS webdriver, then we can either reopen this PR and fix the issues or do a new PR.

**Pull request closed without merging**